### PR TITLE
GH1609: Add additional VSTS actions

### PR DESCRIPTION
**Pull request by louisfischer**
This pull request added  PublishTestResults and PublishCodeCoverage commands to the TFSBuilder. This is related to open issue #1609

### CODE REVIEW

**Comment by louisfischer** [comment]
@gep13 or cake team.. Can you see why the build for this pull request is failing on OSX Cake and Ubuntu Cake on Bitwise? I cannot see what the error is.

**Comment by devlead** [comment]
Looks like there's some tests failing 
```
[xUnit.net 00:00:03.8687354]     Cake.Common.Tests.Unit.Build.TFBuild.TFBuildCommandTests+TheCommands.Should_Publish_Code_Coverage [FAIL]
[xUnit.net 00:00:03.8717692]       Assert.Contains() Failure
[xUnit.net 00:00:03.8724098]       Not found: (filter expression)
[xUnit.net 00:00:03.8729354]       In value:  List<FakeLogMessage> [FakeLogMessage { Level = Information, Message = "##vso[codecoverage.publish codecoveragetool=Cobert"..., Verbosity = Quiet }]
Failed   Cake.Common.Tests.Unit.Build.TFBuild.TFBuildCommandTests+TheCommands.Should_Publish_Code_Coverage
Error Message:
 Assert.Contains() Failure
Not found: (filter expression)
In value:  List<FakeLogMessage> [FakeLogMessage { Level = Information, Message = "##vso[codecoverage.publish codecoveragetool=Cobert"..., Verbosity = Quiet }]
[xUnit.net 00:00:03.8884922]     Cake.Common.Tests.Unit.Build.TFBuild.TFBuildCommandTests+TheCommands.Should_Publish_Test_Results [FAIL]
[xUnit.net 00:00:03.8886023]       Assert.Contains() Failure
[xUnit.net 00:00:03.8886587]       Not found: (filter expression)
[xUnit.net 00:00:03.8887030]       In value:  List<FakeLogMessage> [FakeLogMessage { Level = Information, Message = "##vso[results.publish type=XUnit;mergeResults=true"..., Verbosity = Quiet }]
Failed   Cake.Common.Tests.Unit.Build.TFBuild.TFBuildCommandTests+TheCommands.Should_Publish_Test_Results
Error Message:
 Assert.Contains() Failure
Not found: (filter expression)
In value:  List<FakeLogMessage> [FakeLogMessage { Level = Information, Message = "##vso[results.publish type=XUnit;mergeResults=true"..., Verbosity = Quiet }]
```

File: src/Cake.Common/Build/TFBuild/Data/TFBuildPublishTestResultsData.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            {
+                properties.Add("runTitle", $"'{TestRunTitle}'");
+            }
+            if (PublishRunAttachments.HasValue)
+            {
+                properties.Add("publishRunAttachments", PublishRunAttachments.ToString().ToLowerInvariant());
+            }
+            if (TestResultsFiles != null && TestResultsFiles.Any())
+            {
+                properties.Add("resultFiles", string.Join(";", TestResultsFiles.Select(filePath => new FilePath(filePath).MakeAbsolute(environment).FullPath.Replace("/", "\\"))));
```
Comment (jnm2) [inline src/Cake.Common/Build/TFBuild/Data/TFBuildPublishTestResultsData.cs]: @louisfischer Trying to educate myself- how did you find out that it should be semicolon-separated?

Comment (louisfischer) [reply src/Cake.Common/Build/TFBuild/Data/TFBuildPublishTestResultsData.cs]: @jnm2 thanks for that. That is a typo on my part. It should actually be a comma. Turning On System.Debug Variable shows the Logging Commands used by the various tasks built into TFS/VSTS.  

The snippet below (Publish Test Results)  just shows that even the canned TFS tasks use the Logging Commands.

```cmd
adjustedPattern: 'E:\tfsagents\agent4\_work\1\s\test/results/*.xml'
2 matches
2 final results
Reading test results from file 'E:\tfsagents\agent4\_work\1\s\test\results\TestResults.xml'
Reading test results from file 'E:\tfsagents\agent4\_work\1\s\test\results\TestResults2.xml'
Processed: ##vso[results.publish type=NUnit;mergeResults=false;runTitle=17179;publishRunAttachments=true;resultFiles=E:\tfsagents\agent4\_work\1\s\test\results\TestResults.xml,E:\tfsagents\agent4\_work\1\s\test\results\TestResults2.xml;]
task result: Succeeded
Processed: ##vso[task.complete result=Succeeded;]
```
I will be updating the PR for with the right separator.

 I am not sure why the PublishTestResults and PublishCodeCoverage results are not on the [Logging Commands](https://github.com/Microsoft/vsts-tasks/blob/master/docs/authoring/commands.md) documentation page.

@gep13 I read your question to the TFS team that posed a question about this. Did you get any response from them? I was wondering if they would take a pull request to update the documentation. I would be glad to update the docs. I am just not sure if a similar request has already been submitted.

**Comment by devlead** [comment]
@louisfischer your changes have been merged, thanks for your contribution 👍

File: src/Cake.Common/Build/TFBuild/Data/TFBuildPublishTestResultsData.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        /// <summary>
+        /// Gets or Sets the type test runner the results are formated in
+        /// </summary>
+        public TFTestRunnerType? TestRunner { get; set; }
+
+        /// <summary>
+        /// Gets or sets the list of Test Result files to publish.
+        /// </summary>
+        public string[] TestResultsFiles { get; set; }
```
Comment (jnm2) [inline src/Cake.Common/Build/TFBuild/Data/TFBuildPublishTestResultsData.cs]: Any thoughts about making this `IReadOnlyList<string>`? I'm doing `.ToArray()` in my Cake script but I wouldn't have to.

**Pull request merged**